pkg/observable/notifiable/observable_test.go

```diff
@@ -0,0 +1,162 @@
+package notifiable_test
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"golang.org/x/sync/errgroup"
+
+	"pocket/pkg/observable"
+	"pocket/pkg/observable/notifiable"
+)
+
+const (
+	notifyTimeout            = 100 * time.Millisecond
+	unsubscribeSleepDuration = notifyTimeout * 2
+)
+
+func TestNewNotifiableObservable(t *testing.T) {
+	type test struct {
+		name     string
+		notifier chan int
+	}
+
+	input := 123
+	nonEmptyBufferedNotifier := make(chan int, 1)
+	nonEmptyBufferedNotifier <- input
+
+	tests := []test{
+		{name: "nil notifier", notifier: nil},
+		{name: "empty non-buffered notifier", notifier: make(chan int)},
+		{name: "empty buffered len 1 notifier", notifier: make(chan int, 1)},
+		{name: "non-empty buffered notifier", notifier: nonEmptyBufferedNotifier},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx := context.Background()
+			notifee, notifier := notifiable.NewObservable[int](tt.notifier)
+			require.NotNil(t, notifee)
+			require.NotNil(t, notifier)
+
+			// construct 3 distinct subscriptions, each with its own channel
+			subscriptions := make([]observable.Subscription[int], 3)
+			for i := range subscriptions {
+				subscriptions[i] = notifee.Subscribe(ctx)
+			}
+
+			group := errgroup.Group{}
+			notifiedOrTimedOut := func(subscriptionCh <-chan int) func() error {
+				return func() error {
+					// subscriptionCh should receive notified input
+					select {
+					case output := <-subscriptionCh:
+						require.Equal(t, input, output)
+					case <-time.After(notifyTimeout):
+						return fmt.Errorf("timed out waiting for subscription to be notified")
+					}
+					return nil
+				}
+			}
+
+			// ensure all subscription channels are notified
+			for _, subscription := range subscriptions {
+				// concurrently await notification or timeout to avoid blocking on
+				// empty and/or non-buffered notifiers.
+				group.Go(notifiedOrTimedOut(subscription.Ch()))
+			}
+
+			// notify with test input
+			notifier <- input
+
+			// wait for notifee to be notified or timeout
+			err := group.Wait()
+			require.NoError(t, err)
+
+			// unsubscribing should close subscription channel(s)
+			for _, subscription := range subscriptions {
+				subscription.Unsubscribe()
+
+				select {
+				case <-subscription.Ch():
+				default:
+					t.Fatal("subscription channel left open")
+				}
+			}
+		})
+	}
+}
+
+func TestSubscription_Unsubscribe(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	notifee, notifier := notifiable.NewObservable[int](nil)
+	require.NotNil(t, notifee)
+	require.NotNil(t, notifier)
+
+	tests := []struct {
```

Comment:
NIT: You're inline the struct type and initialization here but separating the two above. I think we should pick one (either) and be consistent.

My personal preference is to separate the two but don't feel strongly either way.